docs/design/ovs-pipeline.md

```diff
@@ -333,58 +382,195 @@ more information on connection tracking in OVS.
+2. table=40, priority=0 actions=load:0x1->NXM_NX_REG4[16..18]
+```
+
+Note that, flow 1 (the generated learned flow) has higher priority than flow 2 in table
+[SessionAffinityTable]. When the same client accesses the ClusterIP again, the first packet
+of the connection will be matched by flow 1 due to the match condition `nw_src=10.10.0.2`,
+which is used to match the client, other conditions are used to match the Service.
+
+The actions of flow 1:
+
+* `load:0xa0a0004->NXM_NX_REG3[]` is used to load Endpoint IP to NXM_NX_REG3.
+* `load:0x50->NXM_NX_REG4[0..15]` is used to load Endpoint port number to bits [0..15] in
+  NXM_NX_REG4.
+* `load:0x2->NXM_NX_REG4[16..18]` is used to set value of bits [16..18] in NXM_NX_REG4 to
+  0b010, which indicates that the Service has done Endpoint selection.
+* `load:0x1->NXM_NX_REG0[18]` is used to set bit 18 in NXM_NX_REG4 to 0x1, which indicates
+  that the source and destination MACs should be rewritten.
+
+Note that, as value of bits [16..18] in NXM_NX_REG4 is 0b010, the packet will not be matched
+by any flows in table [ServiceLBTable] except the default one. The default one just forwards
+the packet to the table [EndpointDNATTable]. The client will always access the same Endpoint
+within the session timeout setting by `service.spec.sessionAffinityConfig.clientIP.timeoutSeconds`.
+
+### EndpointDNATTable (42)
+
+The table implements DNAT for Service traffic after Endpoint selection for the first
+packet of Service.
+
+If you dump the flows for this table, you should see flows like the following:
+
+```text
+1. table=42, priority=200,tcp,reg3=0xc0a84d64,reg4=0x2192b/0x7ffff actions=ct(commit,table=45,zone=65520,nat(dst=192.168.77.100:6443),exec(load:0x21->NXM_NX_CT_MARK[]))
+2. table=42, priority=200,tcp,reg3=0xc0a84d65,reg4=0x2286d/0x7ffff actions=ct(commit,table=45,zone=65520,nat(dst=192.168.77.101:10349),exec(load:0x21->NXM_NX_CT_MARK[]))
+3. table=42, priority=200,tcp,reg3=0xa0a0004,reg4=0x20050/0x7ffff actions=ct(commit,table=45,zone=65520,nat(dst=10.10.0.4:80),exec(load:0x21->NXM_NX_CT_MARK[]))
+4. table=42, priority=200,tcp,reg3=0xa0a0102,reg4=0x20050/0x7ffff actions=ct(commit,table=45,zone=65520,nat(dst=10.10.1.2:80),exec(load:0x21->NXM_NX_CT_MARK[]))
+5. table=42, priority=200,udp,reg3=0xa0a0002,reg4=0x20035/0x7ffff actions=ct(commit,table=45,zone=65520,nat(dst=10.10.0.2:53),exec(load:0x21->NXM_NX_CT_MARK[]))
+6. table=42, priority=190,reg4=0x20000/0x70000 actions=load:0x1->NXM_NX_REG4[16..18],resubmit(,41)
+7. table=42, priority=0 actions=resubmit(,45)
+```
+
+For flow 1-5, DNAT is performed with the IP stored in NXM_NX_REG3 and port stored in bits
+[0..15] in NXM_NX_REG4 by `ct commit` action. Note that, the match condition `reg4=0x2192b/0x7ffff`
+is a union value. Value of bits [0..15] is the port number. Value of bits [16..18] is 0b010,
+which indicates that the Service has done Endpoint selection. The zone number is `65520`,
```

Comment:
I feel we can remove the sentences from "The zone number is...".

Reply:
Removed.